bspump/kafka/source.py

```diff
@@ -123,49 +158,78 @@ def __init__(self, app, pipeline, connection, id=None, config=None):
 		self.RefreshTopics = int(self.Config["refresh_topics"])
 		self.LastRefreshTopicsTime = self.App.time()
 
+		# To run the poll on a separate thread
+		self.ProactorService = app.get_service("asab.ProactorService")
+		self.Buffer = []
+		self.BufferLock = asyncio.Lock()
+		self.LastFlushTime = self.App.time()
+		self.Loop = asyncio.get_event_loop()
 
-	async def main(self):
 
+	def poll_kafka(self, consumer):
+		"""
+		This method runs on a thread and fills the buffer with the polled messages from the Kafka consumer.
+
+		Running Kafka's poll method on a separate thread in an asynchronous application enhances responsiveness
+		by freeing the main thread for other tasks. It ensures non-blocking processing, improving throughput and
+		scalability, as message fetching and processing are decoupled.
+		This approach also optimizes resource utilization by leveraging multiple CPU cores and enhances error
+		handling, making the application more resilient and maintainable.
+		"""
 		while self.Running:
+			current_time = self.App.time()
 
-			try:
-				c = confluent_kafka.Consumer(self.ConsumerConfig, logger=L)
+			if self.RefreshTopics > 0 and current_time > self.LastRefreshTopicsTime + self.RefreshTopics:
+				L.info("Topics refreshed in '{}'.".format(self.Id))
+				consumer.unsubscribe()
+				self.LastRefreshTopicsTime = current_time
+				return
 
-			except BaseException as e:
-				L.exception("Error when connecting to Kafka")
-				self.Pipeline.set_error(None, None, e)
+			m = consumer.poll(self.PollInterval)
+
+			if m is None:
+				continue
+
+			if m.error():
+				L.error("The following error occurred while polling for messages: '{}'.".format(m.error()))
+				consumer.unsubscribe()
 				return
 
-			c.subscribe(self.Subscribe)
+			self.Buffer.append(m)
 
-			try:
+			if len(self.Buffer) >= self.BufferSize or (current_time - self.LastFlushTime) > self.BufferTimeout:
 
-				while 1:
-					await self.Pipeline.ready()
-					current_time = self.App.time()
+				try:
+					future = asyncio.run_coroutine_threadsafe(self.flush_buffer(consumer), self.Loop)
+					# Wait for the result:
+					future.result()
```

Comment:
This looks strange, honestly.

Reply:
I know, but this is how it works and they mention it in their official docs: https://docs.python.org/3/library/asyncio-dev.html